### PR TITLE
Add ready flag to ClusterInformation.

### PR DESCRIPTION
**Pull request by fasaxc**
## Description

@robbrockbank Does this look vaguely sane?

## Todos
- [x] Check whether changing the type of the ready flag is an issue for Typha/Felix upgrade
- [x] Tests
- [ ] Documentation
- [x] Release note

## Release Note

```release-note
None required
```

### CODE REVIEW

File: lib/backend/syncersv1/updateprocessors/clusterinfoprocessor.go [hunk 1]
```diff
@@ -27,8 +27,20 @@ func NewClusterInfoUpdateProcessor() watchersyncer.SyncerUpdateProcessor {
 	return NewConfigUpdateProcessor(
 		reflect.TypeOf(apiv2.ClusterInformationSpec{}),
 		DisallowAnnotations,
-		func(node, name string) model.Key { return model.HostConfigKey{Hostname: node, Name: name} },
-		func(name string) model.Key { return model.GlobalConfigKey{Name: name} },
+		func(node, name string) model.Key {
+			if name == "DatastoreReady" {
+				return model.HostReadyFlagKey{
```
Comment (robbrockbank) [inline lib/backend/syncersv1/updateprocessors/clusterinfoprocessor.go]: Do you think it's worth having the Host ready flag? or should we just handle this returning nil?

File: lib/clientv2/client.go [hunk 2]
```diff
@@ -153,12 +165,134 @@ func (p poolAccessor) GetEnabledPools(ipVersion int) ([]net.IPNet, error) {
 // Most Calico deployment scenarios will automatically implicitly invoke this
 // method and so a general consumer of this API can assume that the datastore
 // is already initialized.
-func (c client) EnsureInitialized() error {
+func (c client) EnsureInitialized(ctx context.Context, calicoVersion, clusterType string) error {
```
Comment (robbrockbank) [inline lib/clientv2/client.go]: Looks like we're missing a UT for this.

File: lib/clientv2/client.go [hunk 3]
```diff
@@ -153,12 +165,134 @@ func (p poolAccessor) GetEnabledPools(ipVersion int) ([]net.IPNet, error) {
...
+	if err := c.ensureClusterInformation(ctx, calicoVersion, clusterType); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+const globalClusterInfoName = "default"
+
+// ensureClusterInformation ensures that the ClusterInformation fields i.e. ClusterType,
```
Comment (robbrockbank) [inline lib/clientv2/client.go]: Given that we have this now, should we be calling this from the Node create method to reinstate the HostEndpoint use case.

**Comment by robbrockbank** [comment]
@fasaxc Looks sane to me - couple of comments inline.

File: lib/backend/syncersv1/updateprocessors/clusterinfoprocessor.go (hunk 1 above)
Comment (fasaxc) [reply lib/backend/syncersv1/updateprocessors/clusterinfoprocessor.go]: Think I prefer nil, let's do that.

File: lib/backend/syncersv1/updateprocessors/configurationprocessor.go [hunk 4]
```diff
@@ -61,7 +61,7 @@ func NewConfigUpdateProcessor(
 	allowAnnotations bool,
 	nodeConfigKeyFn NodeConfigKeyFn,
 	globalConfigKeyFn GlobalConfigKeyFn,
-	stringifiers map[string]ValueToStringFn,
+	stringifiers map[string]ConfigFieldValueToV1ModelValue,
```
Comment (robbrockbank) [inline lib/backend/syncersv1/updateprocessors/configurationprocessor.go]: stringifiers -> valueConverters

**Pull request merged**